presto-main/src/main/java/com/facebook/presto/operator/scalar/MergingDigest.java

```diff
@@ -0,0 +1,799 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.operator.scalar;
+
+import com.google.common.base.Preconditions;
+import io.airlift.slice.BasicSliceInput;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+
+import java.util.AbstractCollection;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Maintains a t-digest by collecting new points in a buffer that is then sorted occasionally and merged
+ * into a sorted array that contains previously computed centroids.
+ * <p>
+ * This can be very fast because the cost of sorting and merging is amortized over several insertion. If
+ * we keep N centroids total and have the input array is k long, then the amortized cost is something like
+ * <p>
+ * N/k + log k
+ * <p>
+ * These costs even out when N/k = log k.  Balancing costs is often a good place to start in optimizing an
+ * algorithm.
+ * The virtues of this kind of t-digest implementation include:
+ * <ul>
+ * <li>No allocation is required after initialization</li>
+ * <li>The data structure automatically compresses existing centroids when possible</li>
+ * <li>No Java object overhead is incurred for centroids since data is kept in primitive arrays</li>
+ * </ul>
+ * <p>
+ */
+
+public class MergingDigest
+        extends TDigest
+{
+    private int mergeCount;
+
+    private final double publicCompression;
+    private final double compression;
+
+    // points to the first unused centroid
+    private int activeCentroids;
+
+    private double totalWeight;
+
+    private final double[] weight;
+
+    private final double[] mean;
+
+    private List<List<Double>> data;
+
+    private double unmergedWeight;
+
+    // this is the index of the next temporary centroid
+    // this is a more Java-like convention than activeCentroids uses
+    private int tempUsed;
+    private final double[] tempWeight;
+    private final double[] tempMean;
+    private List<List<Double>> tempData;
+
+    // array used for sorting the temp centroids
+    // to avoid allocations during operation
+    private final int[] order;
+
+    // if true, alternate upward and downward merge passes
+    public boolean useAlternatingSort = true;
```

Comment:
These two flags we should remove and remove any conditional logic associated with them